api/API_VERSIONING.md

```diff
@@ -0,0 +1,203 @@
+namespaces. When we talk about the `vN xDS API`, we really refer to the `N` of the root
+configuration resources (e.g. bootstrap, xDS resources such as `Cluster`). Even though
+`envoy.config.bootstrap.v3.Boostrap` might transitively reference `envoy.service.trace.v2`, this is
+the Envoy static configuration for the v3 API.
+
+# Backwards compatibility
+
+In general, within a package's major API version, we do not allow any breaking changes. The guiding
+principle is that neither the wire format nor protobuf compiler generated language bindings should
+experience a backward compatible break on a change. Specifically:
+
+* Fields should not be renumbered or have their types changed. This is standard proto development
+  procedure.
+
+* Renaming of fields or package namespaces for a proto must not occur. This is inherently dangerous,
+  since:
+  * Fields renames break wire compatibility. This is stricter than standard proto development
+    procedure in the sense that it does not break binary wire format. However, it **does** break
+    loading of YAML/JSON into protos as well as text protos. Since we consider YAML/JSON to be first
+    class inputs, we must not change field names.
+
+  * For service definitions, the gRPC endpoint URL is inferred from package namespace, so this will
+    break client/server communication.
+
+  * For a message embedded in an `Any` object, the type URL, which the package namespace is a part
+    of, may be used by Envoy or other API consuming code. Currently, this applies to the top-level
+    resources embedded in `DiscoveryResponse` objects, e.g. `Cluster`, `Listener`, etc.
+
+  * Consuming code will break and require source code changes to match the API changes.
+
+* Some other changes are considered breaking for Envoy APIs that are usually considered safe in
+  terms of protobuf wire compatibility:
+  * Upgrading a singleton field to a repeated, e.g. `uint32 foo = 1;` to `repeated uint32 foo = 1`.
+    This changes the JSON wire representation and hence is considered a breaking change.
+  
+  * Wrapping an existing field with `oneof`. This has no protobuf or JSON/YAML wire implications,
+    but is disruptive to various consuming stubs in languages such as Go, creating unnecessary
+    churn.
+
+  * Increasing the strictness of
+    [protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate) annotations. Exceptions
+    may be granted for scenarios in which these stricter conditions model behavior already implied
+    structurally or by documentation.
+
+The exception to the above policy is for API versions tagged `vNalpha`. Within an alpha major
+version, arbitrary breaking changes are allowed. Also, within a `vN` major version, messages
+annotated as draft or experimental are exempt from this policy. The following annotations may be
+applied to a protobuf file, message, enum, field or enum value:
+* `[#proto-status: experimental]` indicating that backwards incompatible changes are likely and
```

Comment:
TBH I'm not a fan of these text tags. Can we just kill them and just use alpha to connote this? It seems more straightforward to me.

Reply:
I think this will be less likely to work under the new proposed scheme, in which `vNalpha` is machine generated from the current stable version, and is always just the mechanical transform of the current stable version. Or are you suggesting a `[#alpha:]` tag to simplify all of this?

Reply:
Hmm. Don't we still want the ability to have alpha protos that we are working on even in the context of the major version? Like config dump, etc.?

Reply:
It's an option to discuss. My current thinking is that we should ideally have the "next release candidate" be somehow automatically generated from the current stable version, applying the intended deprecations and mechanical changes.

For brand new packages, I think we can do manual alpha, like we have today. For existing packages, I think if we adopt the above philosophy and annotations, we're in a much more maintainable situation. A single source of truth, a single place to do active development (modulo the allowances for additions to the previous stable version), etc.

Reply:
I agree with the above, but if we can put new packages in an alpha namespace, I'm mainly wondering if we need this comment annotation at all? Meaning, I feel like a field in a stable proto shouldn't really be experimental, and if it points to an alpha proto it's inherently experimental?

Reply:
After doing a brief survey, I agree. I'm going to go in and remove all these annotations from the v2[alpha] API. I think the annotations had value before we had a strong story around versioning, now we should be encouraging folks to "do the right thing" and be stable within any given version; they can always deprecate stuff and have it cleaned up at the next API clock tick.